### ISSUE TITLE
Use multiple registry keys for some games

### ISSUE THREAD

**Issue by sibir-ine** [loot/loot#1539]
This was discussed on Discord [here](https://discord.com/channels/473542112974077963/473542230095822848/847167052397412422) & [here](https://discord.com/channels/473542112974077963/473542230095822848/851079389247438868), but the registry keys we use to locate some games are not universal across all retail versions of those games (i.e., Steam, GOG, disc). I think it's unreasonable to accommodate every single retail version of every single game to this extent, but I think we should try to accommodate the most common versions with simple registry keys (i.e., Steam, GOG), even if they use different registry keys, as these differences are brought to our attention. I've noticed this with two games so far: Morrowind & Nehrim.

For Morrowind, we currently use the registry subkey at `HKEY_LOCAL_MACHINE\SOFTWARE\Bethesda Softworks\Morrowind\Installed Path`/`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Bethesda Softworks\Morrowind\Installed Path`. For the GOG version, this subkey exists & is sufficient for LOOT. However, for the Steam version, this subkey doesn't exist even though the key does. There is a subkey at `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Steam App 22320\InstallLocation` that should be able to be used for this purpose as well.

For Nehrim, we currently use the registry subkey at `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Nehrim - At Fate's Edge_is1\InstallLocation`/`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\Nehrim - At Fate's Edge_is1\InstallLocation`. For the non-Steam version available from SureAI's [website](https://sureai.net/games/nehrim/), this subkey exists & is sufficient for LOOT. However, for the Steam version, this key doesn't exist at all. There is a subkey at `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Steam App 1014940\InstallLocation` that should be able to be used for this purpose as well.

We could just add new LOOT settings entries for these (i.e., something like TES III: Morrowind - Steam), but it might be worth allowing entries to search multiple registry keys if it's not too much more complicated.

However, another issue I found that complicates this one is that, from testing, these Steam-exclusive keys don't seem to be able to be used to find the install path of either game. I've tested with both games with both the latest stable & snapshot builds of LOOT as well as with several other games (i.e., FO3, FNV, TES4) & their equivalent keys, & none of them worked. I checked [my debug log](https://github.com/loot/loot/files/6623226/LOOTDebugLog.txt) to make sure they were pointing to the right path, & they were. I noticed that the path values in these subkeys lack a closing backslash where the default subkeys don't, so I even tried manually adding them to see if that helped, but it didn't. Unfortunately, I don't possess the technical knowledge to pursue this issue further to discover whether it's user error, something inherent to the keys that makes them unusable, or something that needs to be changed in LOOT to use them.

Another good reason to pursue this issue is that, from testing, the install paths at these Steam-exclusive keys seem to be persistent through validations & exist before even running a game launcher for the first time. This means that, if they can be used & we add support for multiple keys in a single entry, they can be used in addition to the default keys to mostly eliminate the common report from users that LOOT couldn't find their game when caused by these issues.

I say mostly as I've found one significant limitation to these keys. From testing, if you move a game to another directory through Steam, the subkey isn't updated. You have to completely reinstall the game to update it. Therefore, if we do decide to use these keys in addition to the ones we already use, I think we should set priorities for the keys such that the first key accessed that returns a proper path is used without checking succeeding keys. While this is, to some extent, kicking the can down the road, I feel like far less people will encounter this issue than the previously mentioned ones, & we will only have to worry about it with a handful of games. For the other games where the default keys work for Steam versions but a user has moved a game without reinstalling, we can simply tell them to run the launcher like usual so LOOT can just use the default key.

**Comment by Ortham** [comment]
This is implemented in 3a7cd6853a52a1ef297f40a6c05151547d64fcf0. Aside from the Morrowind and Nehrim keys given above, are there any listed for the other games anywhere?

**Comment by sibir-ine** [comment]
For games installed through Steam, they all have similar keys to the two but with their own Steam App ID (i.e., Morrowind is 22320, Nehrim is 1014940). The IDs for every game can be searched [here](https://steamdb.info/apps/). However, I haven't come across any other cases necessitating multiple keys.

**Comment by Ortham** [comment]
OK, I've got Oblivion on Steam and GOG, so I'll check them. It might be a good idea to list all the Steam registry keys too, if that means people don't need to run their game launcher (which catches quite a few users out), so I'll check some other Steam games too.

**Comment by Ortham** [comment]
The Steam registry keys above don't work in LOOT because LOOT uses the 32-bit Registry view when looking up keys (because all existing keys are stored in that view) but the Steam keys are in the native Registry view, so on 64-bit Windows LOOT can't see them.

I don't want to complicate how keys are configured to say "this is a 32-bit key, this is a native key", so instead I think I'm just going to look it up in the 32-bit view first and then try the native view if that fails. 32-bit first because most lookups will be for the non-Steam keys.

**Comment by Ortham** [comment]
Right, I've added Steam and GOG registry keys for all games (for GOG, all games it sells). The keys are tried in the order they're listed, first in the 32-bit Registry and then in the native (i.e. 64-bit on 64-bit Windows, 32-bit on 32-bit Windows) Registry if the first lookup fails. Once lookup succeeds, if its value isn't a valid game install path (i.e. no game master file in its data directory), LOOT moves on to the next one.

The first key listed for each game is the key added by its launcher, because as Sibir noted above that's most likely to stay accurate. The exception is Nehrim, which has the key added by its installer first, because as far as I'm aware its launcher doesn't add its own key - I can't run the Steam version's launcher because the game requires Steam, Nehrim and Oblivion to all be installed on the same drive, but I install games to a second drive...

Some games have multiple store entries, I've added keys for all the app IDs that steamdb/gogdb know about.
